File: chronos/src/Chronos.ProfilerAgent/Int64Marshaler.cpp
#include "StdAfx.h"
#include "Int64Marshaler.h"


void CInt64Marshaler::Marshal(__long value, CBaseStream* stream)
{
	stream->Write(&value, TypeSize::_INT64);
}

__long CInt64Marshaler::Demarshal(CBaseStream* stream)
{
	__long value = 0;
	stream->Read(&value, TypeSize::_INT64);
	return value;
}

File: Chronos/source/Chronos/Marshaling/Generic/Int64Marshaler.cs
using System.IO;

namespace Chronos.Marshaling
{
    public class Int64Marshaler : GenericMarshaler<long>
    {
        protected override void MarshalInternal(long value, Stream stream)
        {
            Marshal(value, stream);
        }

        protected override long DemarshalInternal(Stream stream)
        {
            return Demarshal(stream);
        }

        public static unsafe void Marshal(long value, Stream stream)
        {
            byte[] buffer = new byte[sizeof (long)];
            fixed (byte* pointer = buffer)
            {
                *((long*) pointer) = value;
            }
            stream.Write(buffer, 0, buffer.Length);
        }

        public static unsafe long Demarshal(Stream stream)
        {
            byte[] buffer = new byte[sizeof (long)];
            stream.Read(buffer, 0, buffer.Length);
            long value;
            fixed (byte* pointer = buffer)
            {
                value = *((long*) pointer);
            }
            return value;
        }
    }
}
